README.md
[![Stories in Ready](https://badge.waffle.io/OpenBCI/OpenBCI_NodeJS.png?label=ready&title=Ready)](https://waffle.io/OpenBCI/OpenBCI_NodeJS)
[![Join the chat at https://gitter.im/OpenBCI/OpenBCI_NodeJS](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/OpenBCI/OpenBCI_NodeJS?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Build Status](https://travis-ci.org/OpenBCI/OpenBCI_NodeJS.svg?branch=master)](https://travis-ci.org/OpenBCI/OpenBCI_NodeJS)
[![codecov](https://codecov.io/gh/OpenBCI/OpenBCI_NodeJS/branch/master/graph/badge.svg)](https://codecov.io/gh/OpenBCI/OpenBCI_NodeJS)
[![Dependency Status](https://david-dm.org/OpenBCI/OpenBCI_NodeJS.svg)](https://david-dm.org/OpenBCI/OpenBCI_NodeJS)
[![npm](https://img.shields.io/npm/dm/openbci.svg?maxAge=2592000)](http://npmjs.com/package/openbci)
[![js-semistandard-style](https://img.shields.io/badge/code%20style-semistandard-brightgreen.svg?style=flat-square)](https://github.com/Flet/semistandard)

# OpenBCI Node.js SDK

A Node.js module for OpenBCI ~ written with love by [Push The World!](http://www.pushtheworldllc.com)

We are proud to support all functionality of the Cyton (8 and 16 Channel boards). For the Ganglion (4 channel) please visit [OpenBCI_NodeJS_Ganglion](https://github.com/OpenBCI/OpenBCI_NodeJS_Ganglion). Push The World is actively developing and maintaining this module.

The purpose of this module is to **get connected** and **start streaming** as fast as possible.

Python researcher or developer? Check out how easy it is to [get access to the entire API in the Python example](examples/python)!

### Table of Contents:
---

1. [TL;DR](#tldr)
2. [Installation](#install)
3. [Cyton (32bit Board)](#cyton)
  1. [About](#about)
  2. [General Overview](#general-overview)
  3. [SDK Reference Guide](#sdk-reference-guide)
    * [Constructor](#constructor)
    * [Methods](#method)
    * [Events](#event)
    * [Constants](#constants)
6. [Interfacing With Other Tools](#interfacing-with-other-tools)
7. [Developing](#developing)
8. [Testing](#developing-testing)
9. [Contribute](#contribute)
10. [License](#license)
11. [Roadmap](#roadmap)

### <a name="install"></a> Installation:
```
npm install openbci
```
#### serialport dependency
If you encounter this error when trying to run:
```
Error: The module '/path/to/your/project/node_modules/serialport/build/Release/serialport.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 48. This version of Node.js requires
NODE_MODULE_VERSION 51. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or`npm install`).
```
...the issue can be resolved by running:
```
npm rebuild --build-from-source
```
### <a name="tldr"></a> TL;DR:
Get connected and [start streaming right now with the example code](examples/getStreaming/getStreaming.js).

#### Cyton (8 and 16 channel boards)
```ecmascript 6
const Cyton = require('openbci').Cyton;
const ourBoard = new Cyton();
ourBoard.connect(portName) // Port name is a serial port name, see `.listPorts()`
  .then(() => {
    ourBoard.on('ready',() => {
      ourBoard.streamStart();
      ourBoard.on('sample',(sample) => {
        /** Work with sample */
        for (let i = 0; i < ourBoard.numberOfChannels(); i++) {
          console.log("Channel " + (i + 1) + ": " + sample.channelData[i].toFixed(8) + " Volts.");
          // prints to the console
          //  "Channel 1: 0.00001987 Volts."
          //  "Channel 2: 0.00002255 Volts."
          //  ...
          //  "Channel 8: -0.00001875 Volts."
          }
        });
    });
});
```

# <a name="cyton"></a> Cyton

## <a name="about"></a> About:
Want to know if the module really works? Check out some projects and organizations using it:

* [_OpenEXP_](https://github.com/openexp/OpenEXP): an open-source desktop app for running experiments and collecting behavioral and physiological data.
* [_Thinker_](http://www.pushtheworldllc.com/): a project building the world's first brainwave-word database.
* [_NeuroJS_](https://github.com/NeuroJS): a community dedicated to Neuroscience research using JavaScript, they have several great examples.

Still not satisfied it works?? Check out this [detailed report](http://s132342840.onlinehome.us/pushtheworld/files/voltageVerificationTestPlanAndResults.pdf) that scientifically validates the output voltages of this module.

How are you still doubting and not using this already? Fine, go look at some of the [800 **_automatic_** tests](https://codecov.io/gh/OpenBCI/OpenBCI_NodeJS) written for it!

## <a name="general-overview"></a> General Overview:

Python researcher or developer? Check out how easy it is to [get access to the entire API in the Python example](examples/python)!

Initialization
--------------

Initializing the board:

```js
const Cyton = require('openbci').Cyton;
const ourBoard = new Cyton();
```
Go [checkout out the get streaming example](examples/getStreaming/getStreaming.js)!

For initializing with options, such as verbose print outs:

```js
const Cyton = require('openbci').Cyton;
const ourBoard = new Cyton({
  verbose: true
});
```

Or if you don't have a board and want to use synthetic data:

```js
const Cyton = require('openbci').Cyton;
const ourBoard = new Cyton({
  simulate: true
});
```

Have a daisy?:
```js
var Cyton = require('openbci').Cyton;
var ourBoard = new Cyton({
    boardType: `daisy`,
    hardSet: true
});
```
Go [checkout out the get streaming with daisy example](examples/getStreamingDaisy/getStreamingDaisy.js)!

Another useful way to start the simulator:
```js
const Cyton = require('openbci').Cyton;
const k = require('openbci-utilities').Constants;
const ourBoard = new Cyton();
ourBoard.connect(k.OBCISimulatorPortName) // This will set `simulate` to true
    .then((boardSerial) => {
        ourBoard.on('ready',() => {
            /** Start streaming, reading registers, what ever your heart desires  */
        });
    }).catch((err) => {
        /** Handle connection errors */
    });
```

or if you are using ES6:
```js
import { Cyton } from 'openbci';
import { Constants } from 'openbci-utilities';
const ourBoard = new Cyton();
ourBoard.connect(Constants.OBCISimulatorPortName);
```

To debug, it's amazing, do:
```js
const Cyton = require('openbci').Cyton;
const ourBoard = new Cyton({
    debug: true
});
```
Go [checkout out the debug example](examples/debug/debug.js)!

'ready' event
------------

You MUST wait for the 'ready' event to be emitted before streaming/talking with the board. The ready happens asynchronously
so installing the 'sample' listener and writing before the ready event might result in... nothing at all.

```js
const Cyton = require('openbci').Cyton;
const ourBoard = new Cyton();
ourBoard.connect(portName).then(function(boardSerial) {
    ourBoard.on('ready',function() {
        /** Start streaming, reading registers, what ever your heart desires  */
    });
}).catch(function(err) {
    /** Handle connection errors */
});            
```

Sample properties:
------------------
* `startByte` (`Number` should be `0xA0`)
* `sampleNumber` (a `Number` between 0-255)
* `channelData` (channel data indexed at 0 filled with floating point `Numbers` in Volts)
* `accelData` (`Array` with X, Y, Z accelerometer values when new data available)
* `auxData` (`Buffer` filled with either 2 bytes (if time synced) or 6 bytes (not time synced))
* `stopByte` (`Number` should be `0xCx` where x is 0-15 in hex)
* `boardTime` (`Number` the raw board time)
* `timeStamp` (`Number` the `boardTime` plus the NTP calculated offset)

The power of this module is in using the sample emitter, to be provided with samples to do with as you wish.

You can also start the simulator by sending [`.connect(portName)`](#method-connect) with `portName` equal to [`'OpenBCISimulator'`](#constants-obcisimulatorportname).

To get a ['sample'](#event-sample) event, you need to:
-------------------------------------
1. Call [`.connect(serialPortName)`](#method-connect)
2. Install the ['ready'](#event-ready) event emitter on resolved promise
3. In callback for ['ready'](#event-ready) emitter, call [`streamStart()`](#method-stream-start)
4. Install the ['sample'](#event-sample) event emitter
```js
const Cyton = require('openbci').Cyton;
const ourBoard = new Cyton();
ourBoard.connect(portName).then(function() {
    ourBoard.on('ready',function() {
        ourBoard.streamStart();
        ourBoard.on('sample',function(sample) {
            /** Work with sample */
        });
    });
}).catch(function(err) {
    /** Handle connection errors */
});            
```
Close the connection with [`.streamStop()`](#method-stream-stop) and disconnect with [`.disconnect()`](#method-disconnect)
```js
const Cyton = require('openbci').Cyton;
const ourBoard = new Cyton();
ourBoard.streamStop().then(ourBoard.disconnect());
```

Time Syncing
------------
You must be using OpenBCI firmware version 2 in order to do time syncing. After you [`.connect()`](#method-connect) and send a [`.softReset()`](#method-soft-reset), you can call [`.usingVersionTwoFirmware()`](#method-using-version-two-firmware) to get a boolean response as to if you are using `v1` or `v2`.

Now using firmware `v2`, the fun begins! We synchronize the Board's clock with the module's time. In firmware `v2` we leverage samples with time stamps and _ACKs_ from the Dongle to form a time synchronization strategy. Time syncing has been verified to +/- 4ms and a test report is on the way. We are still working on the synchronize of this module and an NTP server, this is an open call for any NTP experts out there! With a global NTP server you could use several different devices and all sync to the same time server. That way you can really do some serious cloud computing!

Keep your resync interval above 50ms. While it's important to resync every couple minutes due to drifting of clocks, please do not try to sync without getting the last sync event! We can only support one sync operation at a time!

Using local computer time:
```js
const Cyton = require('openbci').Cyton;
const k = require('openbci-utilities').Constants;
const ourBoard = new Cyton({
  verbose:true
});

const resyncPeriodMin = 5; // re sync every five minutes
const secondsInMinute = 60;
let sampleRate = k.OBCISampleRate250; // Default to 250, ALWAYS verify with a call to `.sampleRate()` after 'ready' event!
let timeSyncPossible = false;

// Call to connect
ourBoard.connect(portName).then(() => {
  ourBoard.on('ready',() => {
    // Get the sample rate after 'ready'
    sampleRate = ourBoard.sampleRate();
    // Find out if you can even time sync, you must be using v2 and this is only accurate after a `.softReset()` call which is called internally on `.connect()`. We parse the `.softReset()` response for the presence of firmware version 2 properties.
    timeSyncPossible = ourBoard.usingVersionTwoFirmware();

    ourBoard.streamStart()
      .then(() => {
        /** Start streaming command sent to board. */
      })
      .catch(err => {
        console.log(`stream start: ${err}`);
      });
  });

  // PTW recommends sample driven  
  ourBoard.on('sample',sample => {
    // Resynchronize every every 5 minutes
    if (sample._count % (sampleRate * resyncPeriodMin * secondsInMinute) === 0) {
      ourBoard.syncClocksFull()
        .then(syncObj => {
          // Sync was successful
          if (syncObj.valid) {
            // Log the object to check it out!
            console.log(`syncObj`,syncObj);
    
            // Sync was not successful
            } else {
              // Retry it
              console.log(`Was not able to sync, please retry?`);
            }
        });
    }
    
    if (sample.timeStamp) { // true after the first sync
      console.log(`NTP Time Stamp ${sample.timeStamp}`);
    }
    
  });
})
.catch(err => {
  console.log(`connect: ${err}`);
});
```

Auto-finding boards
-------------------
You must have the OpenBCI board connected to the PC before trying to automatically find it.

If a port is not automatically found, then call [`.listPorts()`](#method-list-ports) to get a list of all serial ports this would be a good place to present a drop down picker list to the user, so they may manually select the serial port name.

```js
const Cyton = require('openbci').Cyton;
const ourBoard = new Cyton();
ourBoard.autoFindOpenBCIBoard().then(portName => {
    if(portName) {
        /**
        * Connect to the board with portName
        * i.e. ourBoard.connect(portName).....
        */
    } else {
        /**Unable to auto find OpenBCI board*/
    }
});
```

Note: [`.autoFindOpenBCIBoard()`](#method-auto-find-open-bci-board) will return the name of the Simulator if you instantiate with option `simulate: true`.

Auto Test - (Using impedance to determine signal quality)
---------------------------------------------------------
Measuring impedance is a vital tool in ensuring great data is collected.

**_IMPORTANT!_** Measuring impedance takes time, so *only test what you must*


changelog.md
# 3.0.0

This adds ganglion, wifi, and cyton support to the `openbci` main library. 

If you only need cyton over serial and so not want the additional support, please see [`openbci-cyton`](https://www.npmjs.com/package/openbci-cyton).

# 2.2.1

### Enhancements

* Bumped openbci-utilities to v0.2.0 for new accel data support

# 2.2.0

### Bug Fixes

* Calling sdStart and stop would change the writeOutDelay
* Timeout for v1 was set to 10 ms instead of 50 ms.
* Timeout for v2+ was set to 0 ms instead of 10 ms.

# 2.1.4

### Enhancements

* syncRegisterSettings now resolves channel setting object
* Bumped openbci-utilities to v0.1.5

# 2.1.3

### Enhancements

* Can now sync the channel settings from the actual registers on the ADS! Call `.syncRegisterSettings()`
* Bumped openbci-utilities to v0.1.4

# 2.1.2

### Enhancements

* Update openbci-utilities to v0.1.2 to patch for sendCounts.
* Disconnect did not disconnect from serial port if the device was not streaming.

# 2.1.1

### Bug Fixes

* Closes forgot parentheses in `getBoardType()` #152 (thanks @nateGeorge)

### Enhancements

* Add support for v3 cyton firmware.

# 2.1.0

### Breaking changes

* Significantly reduce the properties in `this.info` object to only have firmware version and number of missed packets. Code dependent on this.info should switch to using `numberOfChannels()`, and `getBoardType()` and `sampleRate()` for accurate board info!

### Enhancements

* Fixes for daisy with new board, specifically hardSet.

# 2.0.1

### Bug Fixes

* Debug bytes was set to always print in processBytes

# 2.0.0

### New Features
* `index.js` file allows for ES6 destructing

### Breaking Changes
* Change name of `OpenBCIBoard` to `Cyton` to follow new naming convention.
    Simply change:
    ```ecmascript 6
    const OpenBCIBoard = require('openbci').OpenBCIBoard;
    const ourBoard = new OpenBCIBoard();
    ```
    ```ecmascript 6
    const Cyton = require('openbci').Cyton;
    const ourBoard = new Cyton();
    ```
* Major change to how board is initialized with removal of `factory` paradigm.
* New dependency called `openbci-utilities`.
* Rejections are errors with messages, so check err.message for info on message, don't expect string.

### Bug Fixes
* Documentation error with `testSignal` function.
 
### Enhancements
* Add more tests for public API functions.

# 1.5.2

### Dependency Package Updates
* `performance-now`: from `^0.2.0` to `2.1.0`
* `serialport` - from `4.0.1` to `4.0.7`

### Development Dependency Package Updates
* `bluebird`: from `3.4.6` to `3.5.0`
* `chai-as-promised`: from `^5.2.0` to `^6.0.0`
* `codecov`: from `^1.0.1` to `^2.1.0`
* `semistandard`: from `^9.0.0` to `^10.0.0`
* `sinon`: from `^1.17.2` to `^2.1.0`
* `snazzy`: from `^5.0.0` to `^6.0.0`

# 1.5.1

### New Features
* Add new example for Lab stream layer (#139) thanks @gabrielibagon

### Breaking changes
* Removed `impedanceCalculationForChannel()` and `impedanceCalculationForAllChannels` from `OpenBCISample.js`

### Bug Fixes
* Fixes #28- Impedance not working properly.

# 1.5.0

### New Features
* New simulator option `simulatorDaisyModuleCanBeAttached` - Boolean, deafults to true, allows the simulation of the a hot swapped daisy board or simulates a misinformed module.
* New `EventEmitter` - `hardSet` - for when the module detects the board is not configured as the options for the module intended and tries to save itself. i.e. when the `daisy` option is `true` and a soft reset message is parsed and the module determines that a daisy was not detected, the module will emit `hardSet` then send an attach daisy command to recover. Either `error` will be emitted if unable to attach or `ready` will be emitted if success.
* Add example for streaming with `daisy` and `hardSet`.

### Breaking changes
* `.setInfoForBoardType()` changed to `.overrideInfoForBoardType()` to elevate it's dangerous nature.
* `.setMaxChannels()` changed to `.hardSetBoardType()` and input changed from numerical to string: 8 and 16 to `default` and `daisy` respectively.

### Bug Fixes
* Fixes #131 - 16 chan not working by sending a channel command and parsing the return.
* Fixed bug where end of transmission characters would not be ejected from buffer.
 
### Enhancements
* Separated radio tests from main board test file.
 
# 1.4.4

### New Features
* Set max number of channels for the board to use with `.setMaxChannels()` see readme.md
* Set the core info object that drives the module with `.overrideInfoForBoardType()` see readme.md
* Get info for the core object that drives the module with `.getInfo()` see readme.md

### Work In Progress
* Bug where daisy would sometimes not be recognized which destroyed all data.

# 1.4.3

### New examples
* Add example of node to python

# 1.4.2

### New examples
* Add example of debug
* Add example of get streaming

# 1.4.1

### Bug Fixes
* Fixes bug where extra data after EOT (`$$$`) was dumped by preserving the poriton after the EOT for further decomposition.
* Fixes bug where any calls to channel set would actually break the openBCIUtilities code as the channelSettingsArray contained an undefined.
* Writes promises resolve when they are actually sent over the serial port. 

# 1.4.0

### New Features

* Three new initialization options: `simulatorFragmentation`, `simulatorBufferSize`, and `simulatorLatencyTimer`.  Together, these enable a more _realistic_ serial port simulation, mimicking different potential user computer systems.
* New option `debug` gives a live dump of serial traffic on the console if enabled
* New API function `.isConnected()` to check if communications are active.
* New API function `.isStreaming()` to check if samples are coming from the board.

### Enhancements

* Implement and adapt semi-standard code style. Closes #83
* autoFindOpenBCIBoard now notices and uses the stock dongle on Linux
* 'synced' object now has `error` property, null on good syncs, error description on bad syncs.

### Breaking Changes


openBCISimulator.js
'use strict';
var EventEmitter = require('events').EventEmitter;
var util = require('util');
var stream = require('stream');

var openBCISample = require('./openBCIUtilities');
var k = openBCISample.k;
var now = require('performance-now');

function OpenBCISimulatorFactory () {
  var factory = this;

  var _options = {
    accel: true,
    alpha: true,
    boardFailure: false,
    daisy: false,
    daisyCanBeAttached: true,
    drift: 0,
    firmwareVersion: [k.OBCIFirmwareV1, k.OBCIFirmwareV2],
    fragmentation: [k.OBCISimulatorFragmentationNone, k.OBCISimulatorFragmentationRandom, k.OBCISimulatorFragmentationFullBuffers, k.OBCISimulatorFragmentationOneByOne],
    latencyTime: 16,
    bufferSize: 4096,
    lineNoise: [k.OBCISimulatorLineNoiseHz60, k.OBCISimulatorLineNoiseHz50, k.OBCISimulatorLineNoiseNone],
    sampleRate: 250,
    serialPortFailure: false,
    verbose: false
  };

  function OpenBCISimulator (portName, options) {
    options = (typeof options !== 'function') && options || {};
    var opts = {};

    stream.Stream.call(this);

    /** Configuring Options */
    var o;
    for (o in _options) {
      var userValue = options[o];
      delete options[o];

      if (typeof _options[o] === 'object') {
        // an array specifying a list of choices
        // if the choice is not in the list, the first one is defaulted to

        if (_options[o].indexOf(userValue) !== -1) {
          opts[o] = userValue;
        } else {
          opts[o] = _options[o][0];
        }
      } else {
        // anything else takes the user value if provided, otherwise is a default

        if (userValue !== undefined) {
          opts[o] = userValue;
        } else {
          opts[o] = _options[o];
        }
      }
    }

    for (o in options) throw new Error('"' + o + '" is not a valid option');

    this.options = opts;

    // Bools
    this.connected = false;
    this.sd = {
      active: false,
      startTime: 0
    };
    this.streaming = false;
    this.synced = false;
    this.sendSyncSetPacket = false;
    // Buffers
    this.outputBuffer = new Buffer(this.options.bufferSize);
    this.outputBuffered = 0;
    // Numbers
    this.channelNumber = 1;
    this.hostChannelNumber = this.channelNumber;
    this.pollTime = 80;
    this.sampleNumber = -1; // So the first sample is 0
    // Objects
    this.sampleGenerator = openBCISample.randomSample(k.OBCINumberOfChannelsDefault, this.options.sampleRate, this.options.alpha, this.options.lineNoise);
    this.time = {
      current: 0,
      start: now(),
      loop: null
    };
    // Strings
    this.portName = portName || k.OBCISimulatorPortName;

    // Call 'open'
    if (this.options.verbose) console.log(`Port name: ${portName}`);
    setTimeout(() => {
      this.connected = true;
      this.emit('open');
    }, 200);
  }

  // This allows us to use the emitter class freely outside of the module
  // TODO: upgrade from old-style streams to stream.Duplex or stream.Transform
  util.inherits(OpenBCISimulator, stream.Stream);

  OpenBCISimulator.prototype.flush = function (callback) {
    this.outputBuffered = 0;

    clearTimeout(this.outputLoopHandle);
    this.outputLoopHandle = null;

    if (callback) callback();
  };

  OpenBCISimulator.prototype.isOpen = function () {
    return this.connected;
  };

  // output only size bytes of the output buffer
  OpenBCISimulator.prototype._partialDrain = function (size) {
    if (!this.connected) throw new Error('not connected');

    if (size > this.outputBuffered) size = this.outputBuffered;

    // buffer is copied because presently openBCICyton.js reuses it
    var outBuffer = new Buffer(this.outputBuffer.slice(0, size));

    this.outputBuffer.copy(this.outputBuffer, 0, size, this.outputBuffered);
    this.outputBuffered -= size;

    this.emit('data', outBuffer);
  };

  // queue some data for output and send it out depending on options.fragmentation
  OpenBCISimulator.prototype._output = function (dataBuffer) {
    // drain full buffers until there is no overflow
    while (this.outputBuffered + dataBuffer.length > this.outputBuffer.length) {
      var len = dataBuffer.copy(this.outputBuffer, this.outputBuffered);
      dataBuffer = dataBuffer.slice(len);
      this.outputBuffered += len;

      this._partialDrain(this.outputBuffered);
      this.flush();
    }

    dataBuffer.copy(this.outputBuffer, this.outputBuffered);
    this.outputBuffered += dataBuffer.length;

    if (!this.outputLoopHandle) {
      var latencyTime = this.options.latencyTime;
      if (this.options.fragmentation === k.OBCISimulatorFragmentationOneByOne ||
          this.options.fragmentation === k.OBCISimulatorFragmentationNone) {
        // no need to wait for latencyTime
        // note that this is the only difference between 'none' and 'fullBuffers'
        latencyTime = 0;
      }
      var outputLoop = () => {
        var size;
        switch (this.options.fragmentation) {
          case k.OBCISimulatorFragmentationRandom:
            if (Math.random() < 0.5) {
              // randomly picked to send out a fragment
              size = Math.ceil(Math.random() * Math.max(this.outputBuffered, 62));
              break;
            } // else, randomly picked to send a complete buffer in next block
            /* falls through */
          case k.OBCISimulatorFragmentationFullBuffers:
          case k.OBCISimulatorFragmentationNone:
          case false:
            size = this.outputBuffered;
            break;
          case k.OBCISimulatorFragmentationOneByOne:
            size = 1;
            break;
        }
        this._partialDrain(size);
        if (this.outputBuffered) {
          this.outputLoopHandle = setTimeout(outputLoop, latencyTime);
        } else {
          this.outputLoopHandle = null;
        }
      };
      if (latencyTime === 0) {
        outputLoop();
      } else {
        this.outputLoopHandle = setTimeout(outputLoop, latencyTime);
      }
    }
  };

  OpenBCISimulator.prototype.write = function (data, callback) {
    if (!this.connected) {
      if (callback) callback('Not connected');
      else throw new Error('Not connected!');
      return;
    }

    // TODO: this function assumes a type of Buffer for radio, and a type of String otherwise
    //       FIX THIS it makes it unusable outside the api code
    switch (data[0]) {
      case k.OBCIRadioKey:
        this._processPrivateRadioMessage(data);
        break;
      case k.OBCIStreamStart:
        if (!this.stream) this._startStream();
        this.streaming = true;
        break;
      case k.OBCIStreamStop:
        if (this.stream) clearInterval(this.stream); // Stops the stream
        this.streaming = false;
        break;
      case k.OBCIMiscSoftReset:
        if (this.stream) clearInterval(this.stream);
        this.streaming = false;
        this._output(new Buffer(`OpenBCI V3 Simulator On Board ADS1299 Device ID: 0x12345 ${this.options.daisy ? `On Daisy ADS1299 Device ID: 0xFFFFF\n` : ``} LIS3DH Device ID: 0x38422 ${this.options.firmwareVersion === k.OBCIFirmwareV2 ? `Firmware: v2.0.0\n` : ``}$$$`));
        break;
      case k.OBCISDLogForHour1:
      case k.OBCISDLogForHour2:
      case k.OBCISDLogForHour4:
      case k.OBCISDLogForHour12:
      case k.OBCISDLogForHour24:
      case k.OBCISDLogForMin5:
      case k.OBCISDLogForMin15:
      case k.OBCISDLogForMin30:
      case k.OBCISDLogForSec14:
        // If we are not streaming, then do verbose output
        if (!this.streaming) {
          this._output(new Buffer('Wiring is correct and a card is present.\nCorresponding SD file OBCI_69.TXT\n$$$'));
        }
        this.sd.active = true;
        this.sd.startTime = now();
        break;
      case k.OBCISDLogStop:
        if (!this.streaming) {
          if (this.SDLogActive) {
            this._output(new Buffer(`Total Elapsed Time: ${now() - this.sd.startTime} ms`));
            this._output(new Buffer(`Max write time: ${Math.random() * 500} us`));
            this._output(new Buffer(`Min write time: ${Math.random() * 200} us`));
            this._output(new Buffer(`Overruns: 0`));
            this._printEOT();
          } else {
            this._output(new Buffer('No open file to close\n'));
            this._printEOT();
          }
        }
        this.SDLogActive = false;
        break;
      case k.OBCISyncTimeSet:
        if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
          this.synced = true;
          setTimeout(() => {
            this._output(new Buffer(k.OBCISyncTimeSent));
            this._syncUp();
          }, 10);
        }
        break;
      case k.OBCIChannelMaxNumber8:
        if (this.options.daisy) {
          this.options.daisy = false;
          this._output(new Buffer(k.OBCIChannelMaxNumber8SuccessDaisyRemoved));
          this._printEOT();
        } else {
          this._printEOT();
        }
        break;
      case k.OBCIChannelMaxNumber16:
        if (this.options.daisy) {
          this._output(new Buffer(k.OBCIChannelMaxNumber16DaisyAlreadyAttached));
          this._printEOT();
        } else {
          if (this.options.daisyCanBeAttached) {
            this.options.daisy = true;
            this._output(new Buffer(k.OBCIChannelMaxNumber16DaisyAttached));
            this._printEOT();
          } else {
            this._output(new Buffer(k.OBCIChannelMaxNumber16NoDaisyAttached));
            this._printEOT();
          }
        }
        break;
      default:
        break;
    }

    /** Handle Callback */
    if (callback) {
      callback(null, 'Success!');
    }
  };

  OpenBCISimulator.prototype.drain = function (callback) {
    if (callback) callback();
  };

  OpenBCISimulator.prototype.close = function (callback) {
    if (this.connected) {
      this.flush();

      if (this.stream) clearInterval(this.stream);

      this.connected = false;
      this.emit('close');
      if (callback) callback();
    } else {
      if (callback) callback('not connected');
    }
  };

  OpenBCISimulator.prototype._startStream = function () {
    var intervalInMS = 1000 / this.options.sampleRate;

    if (intervalInMS < 2) intervalInMS = 2;

    var getNewPacket = sampNumber => {
      if (this.options.accel) {
        if (this.synced) {
          if (this.sendSyncSetPacket) {
            this.sendSyncSetPacket = false;
            return openBCISample.convertSampleToPacketAccelTimeSyncSet(this.sampleGenerator(sampNumber), now().toFixed(0));
          } else {
            return openBCISample.convertSampleToPacketAccelTimeSynced(this.sampleGenerator(sampNumber), now().toFixed(0));
          }
        } else {
          return openBCISample.convertSampleToPacketStandard(this.sampleGenerator(sampNumber));
        }
      } else {
        if (this.synced) {
          if (this.sendSyncSetPacket) {
            this.sendSyncSetPacket = false;
            return openBCISample.convertSampleToPacketRawAuxTimeSyncSet(this.sampleGenerator(sampNumber), now().toFixed(0), new Buffer([0, 0, 0, 0, 0, 0]));
          } else {
            return openBCISample.convertSampleToPacketRawAuxTimeSynced(this.sampleGenerator(sampNumber), now().toFixed(0), new Buffer([0, 0, 0, 0, 0, 0]));
          }
        } else {
          return openBCISample.convertSampleToPacketRawAux(this.sampleGenerator(sampNumber), new Buffer([0, 0, 0, 0, 0, 0]));
        }
      }
    };

    this.stream = setInterval(() => {
      this._output(getNewPacket(this.sampleNumber));
      this.sampleNumber++;
    }, intervalInMS);
  };

  OpenBCISimulator.prototype._syncUp = function () {
    setTimeout(() => {
      this.sendSyncSetPacket = true;
    }, 12); // 3 packets later
  };

  OpenBCISimulator.prototype._printEOT = function () {
    this._output(new Buffer('$$$'));
  };

  OpenBCISimulator.prototype._printFailure = function () {
    this._output(new Buffer('Failure: '));
  };

  OpenBCISimulator.prototype._printSuccess = function () {
    this._output(new Buffer('Success: '));
  };

  OpenBCISimulator.prototype._printValidatedCommsTimeout = function () {
    this._printFailure();
    this._output(new Buffer('Communications timeout - Device failed to poll Host'));
    this._printEOT();
  };

  OpenBCISimulator.prototype._processPrivateRadioMessage = function (dataBuffer) {
    switch (dataBuffer[1]) {
      case k.OBCIRadioCmdChannelGet:
        if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
          if (!this.options.boardFailure) {
            this._printSuccess();
            this._output(new Buffer(`Host and Device on Channel Number ${this.channelNumber}`));
            this._output(new Buffer([this.channelNumber]));
            this._printEOT();
          } else if (!this.serialPortFailure) {
            this._printFailure();
            this._output(new Buffer(`Host on Channel Number ${this.channelNumber}`));
            this._output(new Buffer([this.channelNumber]));
            this._printEOT();
          }
        }
        break;
      case k.OBCIRadioCmdChannelSet:
        if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
          if (!this.options.boardFailure) {
            if (dataBuffer[2] <= k.OBCIRadioChannelMax) {
              this.channelNumber = dataBuffer[2];
              this.hostChannelNumber = this.channelNumber;
              this._printSuccess();
              this._output(new Buffer(`Channel Number ${this.channelNumber}`));
              this._output(new Buffer([this.channelNumber]));
              this._printEOT();
            } else {
              this._printFailure();
              this._output(new Buffer('Verify channel number is less than 25'));
              this._printEOT();
            }
          } else if (!this.serialPortFailure) {
            this._printValidatedCommsTimeout();
          }
        }
        break;
      case k.OBCIRadioCmdChannelSetOverride:
        if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
          if (dataBuffer[2] <= k.OBCIRadioChannelMax) {
            if (dataBuffer[2] === this.channelNumber) {
              this.options.boardFailure = false;
            } else {
              this.options.boardFailure = true;
            }
            this.hostChannelNumber = dataBuffer[2];
            this._printSuccess();
            this._output(new Buffer(`Host override - Channel Number ${this.hostChannelNumber}`));
            this._output(new Buffer([this.hostChannelNumber]));
            this._printEOT();
          } else {
            this._printFailure();
            this._output(new Buffer('Verify channel number is less than 25'));
            this._printEOT();
          }
        }
        break;
      case k.OBCIRadioCmdPollTimeGet:
        if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
          if (!this.options.boardFailure) {
            this._printSuccess();
            this._output(new Buffer(`Poll Time ${this.pollTime}`));
            this._output(new Buffer([this.pollTime]));
            this._printEOT();
          } else {
            this._printValidatedCommsTimeout();
          }
        }
        break;
      case k.OBCIRadioCmdPollTimeSet:
        if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
          if (!this.options.boardFailure) {
            this.pollTime = dataBuffer[2];
            this._printSuccess();
            this._output(new Buffer(`Poll Time ${this.pollTime}`));
            this._output(new Buffer([this.pollTime]));
            this._printEOT();
          } else {
            this._printValidatedCommsTimeout();
          }
        }
        break;
      case k.OBCIRadioCmdBaudRateSetDefault:
        if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
          this._printSuccess();
          this._output(new Buffer('Switch your baud rate to 115200'));
          this._output(new Buffer([0x24, 0x24, 0x24, 0xFF])); // The board really does this
        }
        break;
      case k.OBCIRadioCmdBaudRateSetFast:
        if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
          this._printSuccess();
          this._output(new Buffer('Switch your baud rate to 230400'));
          this._output(new Buffer([0x24, 0x24, 0x24, 0xFF])); // The board really does this
        }
        break;
      case k.OBCIRadioCmdSystemStatus:
        if (this.options.firmwareVersion === k.OBCIFirmwareV2) {
          if (!this.options.boardFailure) {
            this._printSuccess();
            this._output(new Buffer('System is Up'));
            this._printEOT();
          } else {
            this._printFailure();
            this._output(new Buffer('System is Down'));
            this._printEOT();
          }
        }
        break;
      default:
        break;
    }
  };

  factory.OpenBCISimulator = OpenBCISimulator;
}

util.inherits(OpenBCISimulatorFactory, EventEmitter);

module.exports = new OpenBCISimulatorFactory();
